### PR TITLE
Add read markers

### PR DESCRIPTION
**Pull request by ylecollen**
Implement the new read marker API

### CODE REVIEW

**Comment by ylecollen** [comment]
Add

-> Room.getReadMarkerEventId() to get the current read marker event id
-> Room.setReadMarkerEventId(String) to update it

Update

-> Room.sendReadReceipt(callback) is replaced by Room.sendReadReceipt();
-> RoomSummary.getLatestReadEventId() is replaced by RoomSummary.getReadReceiptEventId()

File: matrix-sdk/src/main/java/org/matrix/androidsdk/adapters/MessagesAdapter.java [hunk 1]
```diff
@@ -581,6 +587,23 @@ public void setIsPreviewMode(boolean isPreviewMode) {
         mIsPreviewMode = isPreviewMode;
     }
 
+    /**
+     * Set whether we are ine preview mode to show unread messages
+     * @param isUnreadViewMode
```
Comment (ylecollen) [inline matrix-sdk/src/main/java/org/matrix/androidsdk/adapters/MessagesAdapter.java]: in

File: matrix-sdk/src/main/java/org/matrix/androidsdk/data/Room.java [hunk 2]
```diff
@@ -1127,106 +1151,156 @@ public boolean sendReadReceipt(final ApiCallback<Void> aRespCallback) {
...
-            Log.d(LOG_TAG, "## sendReadReceipt(): roomId=" + getRoomId() + " to " + anEvent.eventId);
+        boolean hasUpdate = false;
 
-            // test if the message has already be read
-            if (getDataHandler().getStore().isEventRead(getRoomId(), getDataHandler().getUserId(), anEvent.eventId)) {
-                Log.d(LOG_TAG, "## sendReadReceipt(): the message was already read");
-                return false;
+        String readMarkerEventId = aReadMarkerEventId;
+        if (!TextUtils.isEmpty(aReadMarkerEventId)) {
+            if (aReadMarkerEventId.startsWith(getRoomId())) {
```
Comment (ylecollen) [inline matrix-sdk/src/main/java/org/matrix/androidsdk/data/Room.java]: MXSession.PATTERN_CONTAIN_MATRIX_MESSAGE_IDENTIFIER should be used.
I assume it is used to check if it is not a tmp event.

File: matrix-sdk/src/main/java/org/matrix/androidsdk/data/Room.java [hunk 3]
```diff
@@ -1127,106 +1151,156 @@ public boolean sendReadReceipt(final ApiCallback<Void> aRespCallback) {
...
+                // Read marker cannot be set to room id, ignore the given read marker
```
Comment (ylecollen) [inline matrix-sdk/src/main/java/org/matrix/androidsdk/data/Room.java]: update the comment

File: matrix-sdk/src/main/java/org/matrix/androidsdk/fragments/MatrixMessageListFragment.java [hunk 4]
```diff
@@ -175,6 +173,7 @@
 
     // default preview mode
     public static final String PREVIEW_MODE_READ_ONLY = "PREVIEW_MODE_READ_ONLY";
+    public static final String PREVIEW_MODE_UNREAD_MESSAGE = "PREVIEW_MODE_UNREAD_MESSAGE";
```
Comment (ylecollen) [inline matrix-sdk/src/main/java/org/matrix/androidsdk/fragments/MatrixMessageListFragment.java]: i don't understand the differences between the both preview modes

File: matrix-sdk/src/main/java/org/matrix/androidsdk/fragments/MatrixMessageListFragment.java [hunk 5]
```diff
@@ -234,6 +233,9 @@ public static MatrixMessageListFragment newInstance(String matrixId, String room
     // y pos of the first visible row
     private int mFirstVisibleRowY = UNDEFINED_VIEW_Y_POS;
 
+    // Id of the dummy event that should become the read marker when server returns the real ID
+    private String mFutureReadMarkerEvent;
+
```
Comment (ylecollen) [inline matrix-sdk/src/main/java/org/matrix/androidsdk/fragments/MatrixMessageListFragment.java]: mFutureReadMarkerEventId.

File: matrix-sdk/src/main/java/org/matrix/androidsdk/fragments/MatrixMessageListFragment.java [hunk 6]
```diff
@@ -955,6 +979,18 @@ public void onSuccess(Void info) {
                         @Override
                         public void run() {
                             onMessageSendingSucceeded(event);
+
+                            if (mFutureReadMarkerEvent != null && prevEventId.equals(mFutureReadMarkerEvent)) {
```
Comment (ylecollen) [inline matrix-sdk/src/main/java/org/matrix/androidsdk/fragments/MatrixMessageListFragment.java]: did you test the case in which the event is echoed before the request gets the response ?
i don't see anything on this implementation.

File: matrix-sdk/src/main/java/org/matrix/androidsdk/adapters/MessagesAdapter.java (hunk 1 above)
Comment (ylecollen) [inline matrix-sdk/src/main/java/org/matrix/androidsdk/adapters/MessagesAdapter.java]: mIsUnreadViewMode seems useless by now.
mIsPreviewMode seems enough.

File: matrix-sdk/src/main/java/org/matrix/androidsdk/fragments/MatrixMessageListFragment.java [hunk 7]
```diff
@@ -479,13 +481,20 @@ public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle sa
...
+            final String previewMode = args.getString(ARG_PREVIEW_MODE_ID);
             // the fragment displays the history around a message
             if (!TextUtils.isEmpty(mEventId)) {
                 mEventTimeLine = new EventTimeline(mSession.getDataHandler(), roomId, mEventId);
                 mRoom = mEventTimeLine.getRoom();
+                if (PREVIEW_MODE_UNREAD_MESSAGE.equals(previewMode)){
+                    mAdapter.setIsUnreadViewMode(true);
+                    final View footerView = ((LayoutInflater) getActivity().getSystemService(Context.LAYOUT_INFLATER_SERVICE))
+                            .inflate(R.layout.unread_messages_footer, null, false);
+                    mMessageListView.addFooterView(footerView);
```
Comment (ylecollen) [inline matrix-sdk/src/main/java/org/matrix/androidsdk/fragments/MatrixMessageListFragment.java]: Don't you think it should be done on the application side ?
the unread mode adds only a footer to the listview.
It seems better that the activity / fragment decides to customise the UI itself.
a single Preview mode seems enough

File: matrix-sdk/src/main/java/org/matrix/androidsdk/adapters/MessagesAdapter.java (hunk 1 above)
Comment (sabrinaj) [inline matrix-sdk/src/main/java/org/matrix/androidsdk/adapters/MessagesAdapter.java]: Done

File: matrix-sdk/src/main/java/org/matrix/androidsdk/data/Room.java (hunk 2 above)
Comment (sabrinaj) [inline matrix-sdk/src/main/java/org/matrix/androidsdk/data/Room.java]: Done

File: matrix-sdk/src/main/java/org/matrix/androidsdk/data/Room.java (hunk 3 above)
Comment (sabrinaj) [inline matrix-sdk/src/main/java/org/matrix/androidsdk/data/Room.java]: Done

**Pull request merged**